### PR TITLE
Refactored output and input into more concrete types.; Added S3 writer.

### PR DESCRIPTION
**Pull request by bwplotka**
cc @4n4nd

### CODE REVIEW

**Comment by 4n4nd** [comment]
I was able to test using the remoteread api with the following command:
```
 go run ./cmd/obslytics export --input-config='{"type":"remoteread","endpoint":"http://demo.robustperception.io:9090/api/v1/read","tls_config":{"insecure_skip_verify":true}}'\
      --output-config='{"type":"parquet","path": "output/up.parquet","storage": {"type":"s3", "config": {"bucket": "DH-DEV-DATA-SCIENCE-TUTORIAL", "endpoint":  "s3.xxxxx.com", }}}' --match="up"\
      --resolution=1h --min-time="2020-09-19T23:59:59Z" --max-time="2020-09-20T00:59:59Z"\
      --debug
```
It works :+1: 
But I was not able to test the Thanos StoreAPI method, because I don't have access to a thanos instance.

File: pkg/exporter/parquet/parquet.go [hunk 1]
```diff
@@ -1,50 +1,39 @@
...
+	"github.com/thanos-community/obslytics/pkg/exporter"
 	parquetwriter "github.com/xitongsys/parquet-go-source/writerfile"
 	"github.com/xitongsys/parquet-go/parquet"
 	"github.com/xitongsys/parquet-go/source"
 	"github.com/xitongsys/parquet-go/writer"
 )
 
-// Output implements output.Output interface.
-type Output struct{}
+var _ exporter.Encoder = &Encoder{}
```
Comment (4n4nd) [inline pkg/exporter/parquet/parquet.go:17]: @bwplotka could you please explain the purpose of this line?

File: pkg/exporter/export.go [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+	go func() {
+		defer w.Close()
+		if err := e.enc.Encode(w, df); err != nil {
+			errch <- errors.Wrap(err, "encode")
+			return
+		}
+		errch <- nil
+	}()
+	defer func() {
+		r.Close()
```
Comment (4n4nd) [inline pkg/exporter/export.go:62]: @bwplotka can you explain why the error output is being ignored here?

**Review by 4n4nd — APPROVED** [review]
A couple comments for some code that I do not understand. 
Could you please fix the linting issue?
Other than that, LGTM :+1:

File: pkg/exporter/parquet/parquet.go (hunk 1 above)
Comment (bwplotka) [reply pkg/exporter/parquet/parquet.go:17]: Added comment (:

File: pkg/exporter/export.go (hunk 2 above)
Comment (bwplotka) [reply pkg/exporter/export.go:62]: Added todos

**Comment by bwplotka** [comment]
Thanks @4n4nd for quick review :heart: 

Hope @iNecas you don't mind if we merge so we can move forward.

**Comment by iNecas** [comment]
I sure don't mind: whatever it takes to move this project forward! :+1:

**Pull request merged**